daemon/images/image_pull.go

```diff
@@ -124,3 +155,130 @@ func (i *ImageService) GetRepository(ctx context.Context, ref reference.Named, a
 	}
 	return repository, confirmedV2, lastError
 }
+
+type contentStoreForPull struct {
+	distribution.ContentStore
+	leases leases.Manager
+
+	mu       sync.Mutex
+	digested []digest.Digest
+}
+
+func (c *contentStoreForPull) addDigested(dgst digest.Digest) {
+	c.mu.Lock()
+	c.digested = append(c.digested, dgst)
+	c.mu.Unlock()
+}
+
+func (c *contentStoreForPull) getDigested() []digest.Digest {
+	c.mu.Lock()
+	digested := make([]digest.Digest, len(c.digested))
+	copy(digested, c.digested)
+	c.mu.Unlock()
+	return digested
+}
+
+func (c *contentStoreForPull) Writer(ctx context.Context, opts ...content.WriterOpt) (content.Writer, error) {
+	w, err := c.ContentStore.Writer(ctx, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return &contentWriter{
+		cs:     c,
+		Writer: w,
+	}, nil
+}
+
+type contentWriter struct {
+	cs *contentStoreForPull
+	content.Writer
+}
+
+func (w *contentWriter) Commit(ctx context.Context, size int64, expected digest.Digest, opts ...content.Opt) error {
+	if err := w.Writer.Commit(ctx, size, expected, opts...); err != nil {
+		return err
+	}
+
+	w.cs.addDigested(expected)
+	return nil
+}
+
+func imageKey(id image.ID) string {
+	return "image-" + id.String()
+}
+
+type imageStoreWithLease struct {
+	image.Store
+	leases   leases.Manager
+	injested *contentStoreForPull
+	ctx      context.Context
+}
+
+func (s *imageStoreWithLease) Create(config []byte) (image.ID, error) {
+	id, err := s.Store.Create(config)
+	if err != nil {
+		return "", err
+	}
+
+	return id, s.updateLease(id)
+}
+
+func (s *imageStoreWithLease) Get(id image.ID) (*image.Image, error) {
+	img, err := s.Store.Get(id)
+	if err != nil {
+		return nil, err
+	}
+	return img, s.updateLease(id)
+}
+
+func (s *imageStoreWithLease) Delete(id image.ID) ([]layer.Metadata, error) {
+	if err := s.leases.Delete(s.ctx, leases.Lease{ID: imageKey(id)}); err != nil && !errdefs.IsNotFound(err) {
+		return nil, errors.Wrap(err, "error deleting lease")
+	}
+	return s.Store.Delete(id)
+}
+
+func (s *imageStoreWithLease) updateLease(id image.ID) error {
+	leaseID := imageKey(id)
+	lease, err := s.leases.Create(s.ctx, leases.WithID(leaseID))
+	if err != nil {
+		if !c8derrdefs.IsAlreadyExists(err) {
+			return errors.Wrap(err, "error creating lease")
+		}
+		lease = leases.Lease{ID: leaseID}
+	}
+
+	digested := s.injested.getDigested()
+	resource := leases.Resource{
+		Type: "content",
+	}
+	for _, dgst := range digested {
+		logrus.WithFields(logrus.Fields{
+			"digest": dgst,
+			"lease":  lease.ID,
+		}).Debug("Adding content digest to lease")
+		resource.ID = dgst.String()
+		if err := s.leases.AddResource(s.ctx, lease, resource); err != nil {
+			return errors.Wrapf(err, "error adding content digest to lease: %s", dgst)
+		}
+	}
+	return nil
+}
+
+func tempLease(ctx context.Context, mgr leases.Manager) (context.Context, func(context.Context) error, error) {
```

Comment:
btw, buildkit sets a special label for leases like this https://github.com/moby/buildkit/blob/master/util/leaseutil/manager.go#L34 just in case. Not a requirement but might want to do the same.